### ISSUE TITLE
web-api:  is further language needed describing AppID (aka rpId) usage ?

### ISSUE THREAD

**Issue by equalsJeffH** [w3c/webauthn#8]
<a href="https://github.com/equalsJeffH"><img src="https://avatars.githubusercontent.com/u/226571?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Originally submitted by: [equalsJeffH](https://github.com/equalsJeffH)**, on: _Monday Aug 24, 2015 at 20:07 GMT_

---

in PR #78 "We need to talk about RP IDs (nee App IDs) here", @balfanz notes:  

>  when the calling app is an Android app there is no callerOrigin for the platform to send to the Authenticator.

My main question regarding incorporating PR #78 is:  given web-api is explicitly a "web api" spec, do we really want/need to address the AppID/rpID-for-native-apps use case in the web-api spec ?  i.e., is it appropriate? 

I am thinking that we should perhaps note the AppID/rpID-for-native-apps aspect in a non-normative NOTE in web-api, and add further language to platform-api-reqs.

Thus I do not recommend merging PR #78 to master as yet.   

Plus,  PR #89 would change the markup of PR #78 somewhat. [this is now moot, PR#89 was merged to master on 25-Aug - JeffH]

See also wrt the term "rpId":  #91

**Comment by equalsJeffH** [comment]
<a href="https://github.com/balfanz"><img src="https://avatars.githubusercontent.com/u/398022?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [balfanz](https://github.com/balfanz)**, on: _Tuesday Aug 25, 2015 at 17:36 GMT_

----

The question at hand is: 

> My main question regarding incorporating PR #78 is: given web-api is explicitly a "web api" spec,
> do we really want/need to address the AppID/rpID-for-native-apps use case in the web-api spec ?
> i.e., is it appropriate?

Yes and no. We don't "address" native-app use cases in the web-api doc in the sense that we don't mention native apps, we don't rationalize technical choices by explaining that they're there for the native-app use case, etc.

***But***, we need to make sure that an Authenticator can work, whether it's being addressed by a web app or by a native app. So when a web browser talks to an Authenticator it needs to be the same way as when a native platform talks to an Authenticator. The language in the PR is an attempt to write something down **only for the web platform** that happens to be same thing that will happen on native platforms.

Does that make sense?

**Comment by equalsJeffH** [comment]
<a href="https://github.com/nadalin"><img src="https://avatars.githubusercontent.com/u/7937635?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [nadalin](https://github.com/nadalin)**, on: _Tuesday Aug 25, 2015 at 22:59 GMT_

----

Wouldn't this (native app part) be covered in the platform requirements document ?

**Comment by equalsJeffH** [comment]
<a href="https://github.com/vijaybh"><img src="https://avatars.githubusercontent.com/u/9886086?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [vijaybh](https://github.com/vijaybh)**, on: _Friday Aug 28, 2015 at 07:36 GMT_

----

Would this be less controversial if we just pulled in the procedure for calculating an RP ID into the web API spec? This was what I was attempting to do with the original text around SOP.

Then we could just point the platform API doc at the web API doc, and say that platforms should respect this method of computing RP IDs.

That would also resolve the editorial issue of having a normative section pointing at an informative doc as a normative reference.

**Comment by equalsJeffH** [comment]
<a href="https://github.com/equalsJeffH"><img src="https://avatars.githubusercontent.com/u/226571?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [equalsJeffH](https://github.com/equalsJeffH)**, on: _Friday Aug 28, 2015 at 13:50 GMT_

----

@vijaybh: I overall agree with your proposal above -- i'd wondered about how the same origin policy (SOP) processing steps had migrated web-api -> platform-api-reqs.   I will mock-up the above in the add-rp-ids branch this morning (hope that's ok -- @balfanz ?)

**Comment by equalsJeffH** [comment]
<a href="https://github.com/equalsJeffH"><img src="https://avatars.githubusercontent.com/u/226571?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [equalsJeffH](https://github.com/equalsJeffH)**, on: _Saturday Aug 29, 2015 at 00:57 GMT_

----

I have performed a nominal re-factoring of platform-api-reqs and web-api per our discussion today. Please review the checked-in current state of PR #78 and the add-rp-ids branch. 

Upon thinking further about this, if the fido-2 TWG is not intending to include the -platform-api-requirements doc in the package that is contributed to W3C, then I am presently thinking we ought to copy much of the "RP ID Concept" section in the revised platform-api-reqs doc into the web-api spec. 

Also, I did not at this time incorporate the specific "domain lowering" language VJ had originally written for the getAssertion authnr processing rules because it is essentially captured in the present platform-api-reqs doc. however, if we copy over the "RP ID Concept" material into web-api  I am thinking we ought to incorp that original language because it references into the current HTML5 spec appropriately (AFAICT) and thus may reduce impedance mismatch with browser developer / W3C folk.

that language is..
```
<p>If the domain of <code>callerOrigin</code>, or any of its superdomains, 
does not match the origin that created the key according to the relaxed 
Same Origin Policy algorithm specified in [[!HTML5]] section 5.3.1, 
the authenticator MUST reject the request with an error.</p>		
```

Note:  I couldn't really grok the alg in [[!HTML5]] section 5.3.1 but that may be due to unfamiliaraity with HTML5 notation & concepts.

**Comment by equalsJeffH** [comment]
<a href="https://github.com/equalsJeffH"><img src="https://avatars.githubusercontent.com/u/226571?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [equalsJeffH](https://github.com/equalsJeffH)**, on: _Wednesday Sep 02, 2015 at 01:18 GMT_

----

@vijaybh merged PR #78, so I think this issue is addressed for RD2. However, I do think as I noted above https://github.com/fido-alliance/fido-2-specs/issues/92#issuecomment-135923651 that we need to incorporate more of the "RP ID Concept" material into web-api. This is non-normative expositive language, so I'll leave this issue open and bump it to the ID milestone.

**Comment by equalsJeffH** [comment]
<a href="https://github.com/equalsJeffH"><img src="https://avatars.githubusercontent.com/u/226571?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Original poster: [equalsJeffH](https://github.com/equalsJeffH)**, on: _Monday Dec 07, 2015 at 19:49 GMT_

----

although we cleaned some of this up when crafting the w3c-fido-submission package, this remains an open issue the w3c web authn wg will need to address once it begins work. thus setting milestone to ms-fido-v2.0-w3c

**Comment by equalsJeffH** [comment]
my comment on PR #83 that we are punting over to here..

I am nominally fine with merging these commits to master given that this is a working draft and nothing is cast in stone as yet.

I am somewhat concerned that we are diluting the description/definition of the RP ID approach in terms of how it also applies to native platform APIs and is a cohesive approach which does not obviate the Web Same Origin Policy and does not supplant federated identity approaches. I argue that we (where "we" is some superset containing interested W3C WebAuthn WG participants) need to determine where this overall "RP ID and facet denotation" approach is defined/published (it seems there are nominally valid arguments that the W3C is not the applicable place), and do so, at least as a explanatory document.

**Comment by selfissued** [comment]
Does the "appid" extension at https://w3c.github.io/webauthn/#sctn-appid-extension language address this?  If so, can this issue be closed?

**Comment by equalsJeffH** [comment]
I think this issue is satisfactorily addressed given the present state of the pre-WD-06 spec now that we have merged PR #464, so am closing it. yay :)
